### ISSUE TITLE
Add method to accepts Android's popup "For a better experience, turn on device location..."

### ISSUE THREAD

**Issue by vincent-hoodoo** [leancodepl/patrol#1147]
Hello,

Thanks a again for the great package!

We like that it makes it possible to test permission requests, so we're running it on FTL with the alpha version of the gcloud CLI, which makes it possible run our tests on [devices that don't have all permissions granted](https://cloud.google.com/sdk/gcloud/reference/alpha/firebase/test/android/run#--grant-permissions).

Unfortunately, when the location permission is requested and accepted, Android devices running on FTL display a second popup that prompts the user to allow Google Location Services to be used.

![cxfnS](https://user-images.githubusercontent.com/104820536/228675040-5e3fabf8-3442-4e74-b8ef-16a75a2b55b5.png)

I tried to accept it by calling `$.native.grantPermissionWhenInUse()`, but it doesn't work at the moment.
It makes it impossible to run the rest of our tests, since that popup blocks the UI.

Would it be possible for patrol to support accepting this popup?

**Comment by vincent-hoodoo** [comment]
Edit: I just ran my tests using the default version of the `gcloud` team, without having to request permissions, and the popup is still showing up.
It is actually displayed when the app requests the user's location to be streamed, for example when calling [`geolocator.getPositionStream()`](https://pub.dev/documentation/geolocator/latest/geolocator/Geolocator/getPositionStream.html)

**Comment by bartekpacia** [comment]
Thanks for reporting this - I saw this popup a few times as well. Very annoying indeed.

Try this:

```dart
if (Platform.isAndroid) {
  await $.native.tap(Selector(text: 'OK'));
}
```

**Comment by vincent-hoodoo** [comment]
Great, thanks a lot for your help! I am running the tests now and should be able to let you how it went in a few minutes.

Quick question: when I run it locally, I don't get the popup, so what do you think would be the best condition to test for instead of `Platform.isAndroid`? Is there a way to test if this popup is present?

**Comment by bartekpacia** [comment]
Hmm, I think you're asking for #499, which we're postponing for a bit too long.

One workaround that comes to my mind is to add a dart-define that is only present when running on FTL:

```console
patrol build android --dart-define 'FIREBASE_TEST_LAB=true'
```

and then in the test, you'd do:


```dart
final onFirebaseTestLab = const bool.fromEnvironment(FIREBASE_TEST_LAB);
if (onFirebaseTestLab) {
  await $.native.tap(Selector(text: 'OK')); // or 'NO THANKS', or maybe 'No thanks'?
}
```

Another workaround would be to run [getNativeViews()](https://pub.dev/documentation/patrol/latest/patrol/NativeAutomator/getNativeViews.html) in a loop few times, and if the dialog appears, dismiss it. See a rough example of `getNativeViews` usage [here](https://github.com/leancodepl/patrol/discussions/1071#discussioncomment-5290365).

**Comment by vincent-hoodoo** [comment]
Great, thank you for the tip, the test passed!

I thought about adding an env variable but I would rather stay away from it if possible, if I can keep their amount reduced to what's necessary that would be more maintainable. 

I added this method as an extension to PatrolTester, what do you think? Is the call `Future.delayed` appropriate here?
```dart
extension PatrolTesterExtension on PatrolTester {
  Future<bool> waitForNativeView(
    Selector selector, {
    int maxAttempts = 5,
    Duration attemptDelay = const Duration(seconds: 1),
  }) async {
    var attempt = 0;
    var views = [];
    while (views.isEmpty && attempt < maxAttempts) {
      await Future.delayed(attemptDelay);
      attempt++;
      views = await native.getNativeViews(selector);
    }
    return views.isNotEmpty;
  }
}
```

**Comment by bartekpacia** [comment]
Looks nice, thanks for sharing this snippet here :)

> Is the call `Future.delayed` appropriate here?

I'd say _it depends_. If it works for you, then it's definitely okay. 

The possible problem is that I don't see any calls to `pumpAndSettle()`, which means Flutter side of your app won't be refereshed.

But I wouldn't just replace `Future.delayed(duration)` with `pumpAndSettle(duration)`, because pumpAndSettle will never finish when the app is not in foreground. For more info on this problem, see #1038 where I attempted to solve this.

**Comment by vincent-hoodoo** [comment]
> The possible problem is that I don't see any calls to pumpAndSettle(), which means Flutter side of your app won't be refereshed.

I run all my tests with
```dart
framePolicy: LiveTestWidgetsFlutterBindingFramePolicy.fullyLive,
```

So that wouldn't be a problem in this case, would it?

**Comment by bartekpacia** [comment]
Wow, that's a pro move.

Yeah, in this case it's not a problem.

**Comment by vincent-hoodoo** [comment]
> Wow, that's a pro move.

Just to clarify, this remark is ironical isn't it? :)

I found it solved some issues I had before, so I've kept it since. Do you know what the pros / cons are of using it?

**Comment by bartekpacia** [comment]
> Just to clarify, this remark is ironical isn't it? :)

No, I really do think it's a smart thing to do (though actually I'd try using [LiveTestWidgetsFlutterBindingFramePolicy.benchmarkLive](https://api.flutter.dev/flutter/flutter_test/LiveTestWidgetsFlutterBindingFramePolicy.html#benchmarkLive) instead). This removes the whole burden of caring about pumping-and-settling.

The only con I know is that in some apps, it works flawlessly, but there was that one app whose Patrol tests became a total dumpster fire after I set `benchmarkLive`/`fullyLive` frame policy. So now, the idea of using frame policy is postponed.

**Comment by bartekpacia** [comment]
I'm going to close this issue since the solution is found. We don't plan to add a special method just for dismissing this pesky Google-location dialog, since it can be handled with `$.native.tap()`.

**Comment by vincent-hoodoo** [comment]
Thank you again so much for your help!

I had quickly read the documentation regarding the differences between `benchmarkLive` and `fullyLive`, I've read it again on the page you linked to, and I'll update my project to use it.
